picasso/src/test/java/com/squareup/picasso/SourceMarkerTest.java

```diff
@@ -0,0 +1,251 @@
+/*
+ * Copyright (C) 2013 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.picasso;
+
+import java.io.IOException;
+import java.util.Arrays;
+import okio.Buffer;
+import okio.BufferedSource;
+import org.junit.Test;
+
+import static com.google.common.truth.Truth.assertThat;
+import static org.junit.Assert.fail;
+
+public final class SourceMarkerTest {
+  @Test public void test() throws Exception {
+    SourceMarker marker = new SourceMarker(new Buffer().writeUtf8("ABCDEFGHIJKLMNOPQRSTUVWXYZ"));
+    BufferedSource source = marker.source();
+
+    assertThat(source.readUtf8(3)).isEqualTo("ABC");
+    long pos3 = marker.mark(7); // DEFGHIJ
+    assertThat(source.readUtf8(4)).isEqualTo("DEFG");
+    long pos7 = marker.mark(5); // HIJKL
+    assertThat(source.readUtf8(4)).isEqualTo("HIJK");
+    marker.reset(pos7); // Back to 'H'
+    assertThat(source.readUtf8(3)).isEqualTo("HIJ");
+    marker.reset(pos3); // Back to 'D'
+    assertThat(source.readUtf8(7)).isEqualTo("DEFGHIJ");
+    marker.reset(pos7); // Back to 'H' again.
+    assertThat(source.readUtf8(6)).isEqualTo("HIJKLM");
+    try {
+      marker.reset(pos7);
+      fail();
+    } catch (IOException expected) {
+      assertThat(expected).hasMessageThat()
+          .isEqualTo("cannot reset to 7: out of range");
+    }
+    try {
+      marker.reset(pos3);
+      fail();
+    } catch (IOException expected) {
+      assertThat(expected).hasMessageThat()
+          .isEqualTo("cannot reset to 3: out of range");
+    }
+  }
+
+  @Test public void exceedLimitTest() throws Exception {
+    SourceMarker marker = new SourceMarker(new Buffer().writeUtf8("ABCDEFGHIJKLMNOPQRSTUVWXYZ"));
+    BufferedSource source = marker.source();
+
+    assertThat(source.readUtf8(3)).isEqualTo("ABC");
+    long pos3 = marker.mark(Long.MAX_VALUE); // DEFGHIJ
```

Comment:
is this comment right? copy-paste accident?

Reply:
Yup. Fixed!